libnetdata/url/url.c

```diff
@@ -79,3 +79,128 @@ char *url_decode_r(char *to, char *url, size_t size) {
 
     return to;
 }
+
+/**
+ * Is request complete?
+ *
+ * The URL will be ready top be parsed when the last characters are the string "\r\n\r\n",
+ * this function checks the last characters of the message.
+ *
+ * @param begin is the first character of the sequence to analyse.
+ * @param end is the last character of the sequence
+ * @param length is the length of the total of bytes read, it is not the difference between end and begin.
+ *
+ * @return It returns HTTP_VALIDATION_OK when the message is complete and HTTP_VALIDATION_INCOMPLETE otherwise.
+ */
+inline HTTP_VALIDATION url_is_request_complete(char *begin,char *end,size_t length) {
+    //Message cannot be complete when first and last address are the same
+    if ( begin == end) {
+        return HTTP_VALIDATION_INCOMPLETE;
+    }
+
+    //We always set as begin the first character of the message, but we wanna check
+    //the last characters, so I am adjusting it case it is possible.
+    if ( length > 3  ) {
+        begin = end - 4;
```

Comment:
Logic:
Can we get out of bounds here? If I have `length == 4` doesn't this mean that `begin` is `begin[0]`, `end` is `begin[3]` and `end - 4` is `begin[-1]`?

Reply:
Begin cannot be negative, because this is mermory allocated by the server. So the pointer will be far from the first positions of the memory, because they are used by the processor when the computer starts, but I agree it is safety to have length >= 4!

Reply:
`if (length >= 4) { ` must change to `if (length > 4) {`. Also, this function is called in two places, and the first place that invokes it may have already done `last_pos -= 4;` which should probably be removed from there.

Reply:
I copied and paste from the old algorithm, and I appended wrong the part of the text, I am fixing for the next commit.